File: src/services/email_body_preview.py
"""Bounded email-body previews shared by serve-api read paths."""

from __future__ import annotations

from bs4 import BeautifulSoup, NavigableString, Tag


EMAIL_BODY_PREVIEW_THRESHOLD_BYTES = 256 * 1024
EMAIL_BODY_PREVIEW_CHARS = 64 * 1024
EMAIL_BODY_HTML_SOURCE_CHARS = 128 * 1024


def _append_within_budget(
    output: BeautifulSoup,
    target: Tag,
    source: Tag | BeautifulSoup,
    remaining: list[int],
) -> None:
    for child in source.contents:
        if remaining[0] <= 0:
            break
        if isinstance(child, NavigableString):
            text = str(child)
            target.append(NavigableString(text[: remaining[0]]))
            remaining[0] -= min(len(text), remaining[0])
            continue
        if not isinstance(child, Tag):
            continue
        clone = output.new_tag(child.name, attrs=dict(child.attrs))
        target.append(clone)
        _append_within_budget(output, clone, child, remaining)


def preview_html(html: str) -> str:
    """Repair a bounded prefix, then serialize balanced nodes within 64K text."""
    source = BeautifulSoup(html, "lxml")
    source_root = source.body or source
    output = BeautifulSoup("", "lxml")
    target = output.new_tag("body")
    output.append(target)
    _append_within_budget(output, target, source_root, [EMAIL_BODY_PREVIEW_CHARS])
    return "".join(str(node) for node in target.contents)

File: frontend/src/electron/main/lib/email-body-preview.ts
import { type HTMLElement, parse } from 'node-html-parser'

export const EMAIL_BODY_PREVIEW_THRESHOLD_BYTES = 256 * 1024
export const EMAIL_BODY_PREVIEW_CHARS = 64 * 1024
export const EMAIL_BODY_HTML_SOURCE_CHARS = 128 * 1024

function appendWithinBudget(
  target: HTMLElement,
  source: HTMLElement,
  remaining: { chars: number }
): void {
  for (const child of source.childNodes) {
    if (remaining.chars <= 0) break
    if (child.nodeType === 3) {
      const text = child.rawText
      if (text.length <= remaining.chars) {
        target.append(child.clone())
        remaining.chars -= text.length
      } else {
        target.append(text.slice(0, remaining.chars))
        remaining.chars = 0
      }
      continue
    }
    if (child.nodeType !== 1) continue

    const element = child as HTMLElement
    const clone = element.clone() as HTMLElement
    target.append(clone)
    if (element.isVoidElement) continue
    clone.set_content('')
    appendWithinBudget(clone, element, remaining)
  }
}

/**
 * Produce valid, balanced HTML without slicing through tags. The SQL layer only
 * reads a bounded source prefix; node-html-parser repairs dirty/unclosed email
 * markup, then this serializer keeps complete element boundaries and truncates
 * only an individual text node when it alone crosses the character budget.
 */
export function previewHtml(html: string, maxChars = EMAIL_BODY_PREVIEW_CHARS): string {
  if (html.length <= maxChars) return html
  const source = parse(html)
  const output = parse('')
  appendWithinBudget(output, source, { chars: maxChars })
  return output.innerHTML
}
